[airflow] Avoid deprecated values (AIR302)

uranusjr opened the pull request:
## Summary

Airflow 3.0 removes various deprecated functions, members, modules, and other values. They have been deprecated in 2.x, but the removal causes incompatibilities that we want to detect.

(We are deprecating a lot more things. I want to use this to establish a basic structure so future checks can be submitted more easily.)

Ref: #14626

## Test Plan

A test fixture is included in the PR.

MichaReiser commented on crates/ruff_linter/src/codes.rs:
```diff
@@ -1031,6 +1031,7 @@ pub fn code_to_rule(linter: Linter, code: &str) -> Option<(RuleGroup, Rule)> {
 
         // airflow
         (Airflow, "001") => (RuleGroup::Stable, rules::airflow::rules::AirflowVariableNameTaskIdMismatch),
+        (Airflow, "302") => (RuleGroup::Preview, rules::airflow::rules::AirflowDeprecatedMembers),
```
Can you tell me a bit about your intended rule numbering? It seems you've something in mind because you chose 302 over e.g 301.

Somewhat related. The numpy rule is called `Numpy2Deprecation`. Should we align this rule's name to `Airflow3Deprecation`?

uranusjr replied:
Well simply because I implemented #14581 (which used AIR301) before this. We don’t really have a rule on these; I can switch them around if it is nicer.

(Now that I think of it, should we start with 300 instead…?)

MichaReiser replied:
> these; I can switch them around if it is nicer.

Ah no, it's fine. Either way works for us.

> (Now that I think of it, should we start with 300 instead…?)

We haven't been very consistent about whether we start with 0 or 1 across linters, but I think we've been mostly consistent for a single inter. That's why I would prefer starting with 301, considering that there's already 001.

MichaReiser commented on crates/ruff_linter/src/rules/airflow/rules/deprecated_members.rs:
```diff
@@ -0,0 +1,92 @@
+/// AIR302
+pub(crate) fn deprecated_members(checker: &mut Checker, expr: &Expr) {
+    let Some((deprecated, replacement)) =
+        checker
```
We could consider adding a `Modules::Airflow` to avoid semantic name lookups for users not using airflow. See https://github.com/astral-sh/ruff/blob/928ffd66503759679823c5f346941ef58486766e/crates/ruff_linter/src/rules/numpy/rules/numpy_2_0_deprecation.rs#L162-L164

uranusjr replied:
Done in #14581 instead.

MichaReiser reviewed (approved):
This is great. Thank you. I've two small questions/suggestions.

sbrugman commented on crates/ruff_linter/src/rules/airflow/rules/deprecated_members.rs:
````diff
@@ -0,0 +1,92 @@
+///
+///
+/// yesterday = today - timedelta(days=1)
+/// ```
````
Is there some documentation available already, such as a migration guide?

uranusjr replied:
Unfortunately there’s not. We may create it later, but would it also make sense to use this as the documentation? I can make the description here a bit longer, and the guide can link back to these rules instead.

MichaReiser replied:
It probably "depends," but I have a slight preference toward having this documentation outside of Ruff, at least if it should also cover *why* something has been deprecated and provide guidance with what it should be replaced instead. Otherwise, the "burden" for maintaining the list falls on us and I don't see us having the authority or knowledge to maintain the list.

uranusjr replied:
Got it, that makes sense. We will create a separate migration guide then. This will be done later and we’ll add the link here.

MichaReiser commented:
Feel free to ping me when you rebased the PR and I'll merge it. This is great work.

uranusjr commented:
Alright, I’ve added argument deprecation to AIR302. I also added some structure in the module since we’re probably going to add other kinds of deprecation (like `foo["deprecated"]`).

dhruvmanila commented on crates/ruff_linter/src/rules/airflow/rules/removal_in_3.rs:
```diff
@@ -0,0 +1,137 @@
+
+fn removed_argument(checker: &mut Checker, qualname: &QualifiedName, arguments: &Arguments) {
+    let deprecations = match qualname.segments() {
+        ["airflow", .., "DAG" | "dag"] => vec![DeprecatedArgument(
+            vec!["timetable", "schedule_interval"],
+            Some("schedule"),
+        )],
```
I think this is the same as the existing `AIR301`, right? If so, we should remove the `AIR301` implementation.

uranusjr replied:
The point of `AIR301` is to detect DAGs without a schedule argument, like this

```python
DAG(dag_id="foo")
```

while the rule here is to detect a DAG using a deprecated schedule argument like

```python
DAG(dag_id="foo", schedule_interval="@daily")
```

We shouldn’t remove AIR301 because AIR302 does not catch the error in the first DAG. I think we should also detect `schedule_interval` and `timetable` in AIR301, so the second DAG only emits AIR302 instead of two errors.

uranusjr replied:
Done in the latest commit

MichaReiser commented on crates/ruff_linter/src/rules/airflow/rules/removal_in_3.rs:
```diff
@@ -0,0 +1,137 @@
+
+fn removed_argument(checker: &mut Checker, qualname: &QualifiedName, arguments: &Arguments) {
+    let deprecations = match qualname.segments() {
+        ["airflow", .., "DAG" | "dag"] => vec![DeprecatedArgument(
+            vec!["timetable", "schedule_interval"],
+            Some("schedule"),
+        )],
```
Returning a `vec` of possible deprecations here is suboptimal for performance because it means that Ruff allocates at least two vectors for every `airflow` call expression. 

I think we could do better by changing the API slightly

```rust
fn deprecated_keyword_argument(arguments: &Arguments, argument: &str, replacement: Option<&str>) -> Option<Diagnostic> {
	let keyword = arguments.find_keyword(argument)?;
	
	Some(Diagnostic::new(
        Airflow3Removal {
            deprecated: argument.to_string(),
            replacement: match replacement {
                Some(name) => Replacement::Name(name.to_owned()),
                None => Replacement::None,
            },
        },
        keyword
            .arg
            .as_ref()
            .map_or_else(|| keyword.range(), Ranged::range),
    ));
}
```

and you can then use it like this

```rust
    let deprecations = match qualname.segments() {
        ["airflow", .., "DAG" | "dag"] => {
					checker.diagnostics.extend(removed_argument(arguments, "timetable", Some("schedule")));
					checker.diagnostics.extend(removed_argument(arguments, "schedule_interval", Some("schedule")));
		}
```

uranusjr replied:
Sounds good to me. I was struggling to find a way to do this more efficiently.

uranusjr replied:
Alright this is pretty nice! Need to suppress Clippy since we do want to add more cases here in the future.
